Android 14 fails to parse NSEC records for custom service type in version 0.128.0

Version:  0.128.0

I register a custom service in python like so:

```
info = zeroconf.ServiceInfo(
        "_green._tcp.local.",
        f"{service_name}._green._tcp.local.",
        port=port,
        properties=desc,
        server=f"{hostname}.local.",
    )

    ZC.register_service(info)
```

android 14 throws this error whenever I try to discover it:

```
2023-12-06 15:28:47.651  1674-2241  serviceDiscovery        pid-1674                             E  [MdnsMultinetworkSocketClient] ERROR Failed to read NSEC record from mDNS response.: Failed to read NSEC record from mDNS response.
                                                                                                    android.net.connectivity.com.android.server.connectivity.mdns.MdnsPacket$ParseException: Failed to read NSEC record from mDNS response.
                                                                                                    	at android.net.connectivity.com.android.server.connectivity.mdns.MdnsPacket.parseRecord(MdnsPacket.java:193)
                                                                                                    	at android.net.connectivity.com.android.server.connectivity.mdns.MdnsPacket.parseRecords(MdnsPacket.java:115)
                                                                                                    	at android.net.connectivity.com.android.server.connectivity.mdns.MdnsPacket.parseRecordsSection(MdnsPacket.java:101)
                                                                                                    	at android.net.connectivity.com.android.server.connectivity.mdns.MdnsResponseDecoder.parseResponse(MdnsResponseDecoder.java:100)
                                                                                                    	at android.net.connectivity.com.android.server.connectivity.mdns.MdnsMultinetworkSocketClient.processResponsePacket(MdnsMultinetworkSocketClient.java:239)
                                                                                                    	at android.net.connectivity.com.android.server.connectivity.mdns.MdnsMultinetworkSocketClient.-$$Nest$mprocessResponsePacket(MdnsMultinetworkSocketClient.java:0)
                                                                                                    	at android.net.connectivity.com.android.server.connectivity.mdns.MdnsMultinetworkSocketClient$ReadPacketHandler.handlePacket(MdnsMultinetworkSocketClient.java:151)
                                                                                                    	at android.net.connectivity.com.android.server.connectivity.mdns.MulticastPacketReader.handlePacket(MulticastPacketReader.java:101)
                                                                                                    	at android.net.connectivity.com.android.server.connectivity.mdns.MulticastPacketReader.handlePacket(MulticastPacketReader.java:34)
                                                                                                    	at android.net.connectivity.com.android.net.module.util.FdEventsReader.handleInput(FdEventsReader.java:272)
                                                                                                    	at android.net.connectivity.com.android.net.module.util.FdEventsReader.lambda$createAndRegisterFd$0(FdEventsReader.java:228)
                                                                                                    	at android.net.connectivity.com.android.net.module.util.FdEventsReader.$r8$lambda$xh0yZpbdLvlMgmz3uMcpPMcWNvc(FdEventsReader.java:0)
                                                                                                    	at android.net.connectivity.com.android.net.module.util.FdEventsReader$$ExternalSyntheticLambda0.onFileDescriptorEvents(R8$$SyntheticClass:0)
                                                                                                    	at android.os.MessageQueue.dispatchEvents(MessageQueue.java:293)
                                                                                                    	at android.os.MessageQueue.nativePollOnce(Native Method)
                                                                                                    	at android.os.MessageQueue.next(MessageQueue.java:335)
                                                                                                    	at android.os.Looper.loopOnce(Looper.java:162)
                                                                                                    	at android.os.Looper.loop(Looper.java:294)
                                                                                                    	at android.os.HandlerThread.run(HandlerThread.java:67)
                                                                                                    Caused by: java.io.IOException: Invalid bitmap length: 0
                                                                                                    	at android.net.connectivity.com.android.server.connectivity.mdns.MdnsNsecRecord.readTypes(MdnsNsecRecord.java:82)
                                                                                                    	at android.net.connectivity.com.android.server.connectivity.mdns.MdnsNsecRecord.readData(MdnsNsecRecord.java:66)
                                                                                                    	at android.net.connectivity.com.android.server.connectivity.mdns.MdnsRecord.<init>(MdnsRecord.java:89)
                                                                                                    	at android.net.connectivity.com.android.server.connectivity.mdns.MdnsNsecRecord.<init>(MdnsNsecRecord.java:43)
                                                                                                    	at android.net.connectivity.com.android.server.connectivity.mdns.MdnsPacket.parseRecord(MdnsPacket.java:191)
                                                                                                    	at android.net.connectivity.com.android.server.connectivity.mdns.MdnsPacket.parseRecords(MdnsPacket.java:115) 
                                                                                                    	at android.net.connectivity.com.android.server.connectivity.mdns.MdnsPacket.parseRecordsSection(MdnsPacket.java:101) 
                                                                                                    	at android.net.connectivity.com.android.server.connectivity.mdns.MdnsResponseDecoder.parseResponse(MdnsResponseDecoder.java:100) 
                                                                                                    	at android.net.connectivity.com.android.server.connectivity.mdns.MdnsMultinetworkSocketClient.processResponsePacket(MdnsMultinetworkSocketClient.java:239) 
                                                                                                    	at android.net.connectivity.com.android.server.connectivity.mdns.MdnsMultinetworkSocketClient.-$$Nest$mprocessResponsePacket(MdnsMultinetworkSocketClient.java:0) 
                                                                                                    	at android.net.connectivity.com.android.server.connectivity.mdns.MdnsMultinetworkSocketClient$ReadPacketHandler.handlePacket(MdnsMultinetworkSocketClient.java:151) 
                                                                                                    	at android.net.connectivity.com.android.server.connectivity.mdns.MulticastPacketReader.handlePacket(MulticastPacketReader.java:101) 
                                                                                                    	at android.net.connectivity.com.android.server.connectivity.mdns.MulticastPacketReader.handlePacket(MulticastPacketReader.java:34) 
                                                                                                    	at android.net.connectivity.com.android.net.module.util.FdEventsReader.handleInput(FdEventsReader.java:272) 
                                                                                                    	at android.net.connectivity.com.android.net.module.util.FdEventsReader.lambda$createAndRegisterFd$0(FdEventsReader.java:228) 
                                                                                                    	at android.net.connectivity.com.android.net.module.util.FdEventsReader.$r8$lambda$xh0yZpbdLvlMgmz3uMcpPMcWNvc(FdEventsReader.java:0) 
                                                                                                    	at android.net.connectivity.com.android.net.module.util.FdEventsReader$$ExternalSyntheticLambda0.onFileDescriptorEvents(R8$$SyntheticClass:0) 
                                                                                                    	at android.os.MessageQueue.dispatchEvents(MessageQueue.java:293) 
                                                                                                    	at android.os.MessageQueue.nativePollOnce(Native Method) 
                                                                                                    	at android.os.MessageQueue.next(MessageQueue.java:335) 
                                                                                                    	at android.os.Looper.loopOnce(Looper.java:162) 
                                                                                                    	at android.os.Looper.loop(Looper.java:294) 
                                                                                                    	at android.os.HandlerThread.run(HandlerThread.java:67) 
```

I searched back in the releases to find a version prior to NSEC support (v0.36.1) and when I use that version, the error goes away and my service is discoverable.

--- a/src/zeroconf/_dns.py
+++ b/src/zeroconf/_dns.py
@@ -493,7 +493,8 @@ class DNSNsec(DNSRecord):
             raise ValueError("NSEC must have at least one rdtype")
         out_bytes = bytes(bitmap[0:total_octets])
         out.write_name(self.next_name)
-        out.write_short(len(out_bytes))
+        out._write_byte(0)  # Always window 0
+        out._write_byte(len(out_bytes))
         out.write_string(out_bytes)
 
     def __eq__(self, other: Any) -> bool:

--- a/src/zeroconf/_dns.py
+++ b/src/zeroconf/_dns.py
@@ -485,9 +485,7 @@ class DNSNsec(DNSRecord):
             if rdtype > 255:  # mDNS only supports window 0
                 raise ValueError(f"rdtype {rdtype} is too large for NSEC")
             byte = rdtype // 8
-            octet_pos = byte + 1
-            if octet_pos > total_octets:
-                total_octets = octet_pos
+            total_octets = byte + 1
             bitmap[byte] |= 0x80 >> (rdtype % 8)
         if total_octets == 0:
             # NSEC must have at least one rdtype

--- a/src/zeroconf/_dns.py
+++ b/src/zeroconf/_dns.py
@@ -495,7 +495,7 @@ class DNSNsec(DNSRecord):
             raise ValueError("NSEC must have at least one rdtype")
         out_bytes = bytes(bitmap[0:total_octets])
         out.write_name(self.next_name)
-        out.write_short(0)
+        out.write_short(0)  # Always window 0
         out.write_short(len(out_bytes))
         out.write_string(out_bytes)
 

--- a/src/zeroconf/_dns.py
+++ b/src/zeroconf/_dns.py
@@ -480,13 +480,19 @@ class DNSNsec(DNSRecord):
     def write(self, out: 'DNSOutgoing') -> None:
         """Used in constructing an outgoing packet."""
         bitmap = bytearray(b'\0' * 32)
+        total_octets = 0
         for rdtype in self.rdtypes:
             if rdtype > 255:  # mDNS only supports window 0
-                continue
-            offset = rdtype % 256
-            byte = offset // 8
-            total_octets = byte + 1
-            bitmap[byte] |= 0x80 >> (offset % 8)
+                raise ValueError(f"rdtype {rdtype} is too large for NSEC")
+            byte = rdtype // 8
+            octet_pos = byte + 1
+            if octet_pos > total_octets:
+                total_octets = octet_pos
+            bitmap[byte] |= 0x80 >> (rdtype % 8)
+        if total_octets == 0:
+            # NSEC must have at least one rdtype
+            # Writing an empty bitmap is not allowed
+            raise ValueError("NSEC must have at least one rdtype")
         out_bytes = bytes(bitmap[0:total_octets])
         out.write_name(self.next_name)
         out.write_short(0)

--- a/src/zeroconf/_dns.py
+++ b/src/zeroconf/_dns.py
@@ -493,7 +493,6 @@ class DNSNsec(DNSRecord):
             raise ValueError("NSEC must have at least one rdtype")
         out_bytes = bytes(bitmap[0:total_octets])
         out.write_name(self.next_name)
-        out.write_short(0)  # Always window 0
         out.write_short(len(out_bytes))
         out.write_string(out_bytes)
 

--- a/tests/test_protocol.py
+++ b/tests/test_protocol.py
@@ -12,6 +12,8 @@ import unittest
 import unittest.mock
 from typing import cast
 
+import pytest
+
 import zeroconf as r
 from zeroconf import DNSHinfo, DNSIncoming, DNSText, const, current_time_millis
 
@@ -65,7 +67,7 @@ class PacketGeneration(unittest.TestCase):
         parsed = r.DNSIncoming(generated.packets()[0])
         assert answer in parsed.answers()
 
-        # Types > 255 should be ignored
+        # Types > 255 should raise an exception
         answer_invalid_types = r.DNSNsec(
             'eufy HomeBase2-2464._hap._tcp.local.',
             const._TYPE_NSEC,
@@ -76,8 +78,22 @@ class PacketGeneration(unittest.TestCase):
         )
         generated = r.DNSOutgoing(const._FLAGS_QR_RESPONSE)
         generated.add_answer_at_time(answer_invalid_types, 0)
-        parsed = r.DNSIncoming(generated.packets()[0])
-        assert answer in parsed.answers()
+        with pytest.raises(ValueError, match='rdtype 1000 is too large for NSEC'):
+            generated.packets()
+
+        # Empty rdtypes are not allowed
+        answer_invalid_types = r.DNSNsec(
+            'eufy HomeBase2-2464._hap._tcp.local.',
+            const._TYPE_NSEC,
+            const._CLASS_IN | const._CLASS_UNIQUE,
+            const._DNS_OTHER_TTL,
+            'eufy HomeBase2-2464._hap._tcp.local.',
+            [],
+        )
+        generated = r.DNSOutgoing(const._FLAGS_QR_RESPONSE)
+        generated.add_answer_at_time(answer_invalid_types, 0)
+        with pytest.raises(ValueError, match='NSEC must have at least one rdtype'):
+            generated.packets()
 
     def test_parse_own_packet_response(self):
         generated = r.DNSOutgoing(const._FLAGS_QR_RESPONSE)

--- a/tests/test_protocol.py
+++ b/tests/test_protocol.py
@@ -67,6 +67,21 @@ class PacketGeneration(unittest.TestCase):
         parsed = r.DNSIncoming(generated.packets()[0])
         assert answer in parsed.answers()
 
+        # Now with the higher RD type first
+        answer = r.DNSNsec(
+            'eufy HomeBase2-2464._hap._tcp.local.',
+            const._TYPE_NSEC,
+            const._CLASS_IN | const._CLASS_UNIQUE,
+            const._DNS_OTHER_TTL,
+            'eufy HomeBase2-2464._hap._tcp.local.',
+            [const._TYPE_SRV, const._TYPE_TXT],
+        )
+
+        generated = r.DNSOutgoing(const._FLAGS_QR_RESPONSE)
+        generated.add_answer_at_time(answer, 0)
+        parsed = r.DNSIncoming(generated.packets()[0])
+        assert answer in parsed.answers()
+
         # Types > 255 should raise an exception
         answer_invalid_types = r.DNSNsec(
             'eufy HomeBase2-2464._hap._tcp.local.',

